kani-compiler/src/kani_middle/kani_functions.rs

```diff
@@ -0,0 +1,232 @@
+    type Error = ();
+
+    fn try_from(def: FnDef) -> Result<Self, Self::Error> {
+        let value = attributes::fn_marker(def).ok_or(())?;
+        if let Ok(intrisic) = KaniIntrinsic::from_str(&value) {
+            Ok(intrisic.into())
+        } else if let Ok(model) = KaniModel::from_str(&value) {
+            Ok(model.into())
+        } else if let Ok(hook) = KaniHook::from_str(&value) {
+            Ok(hook.into())
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl TryFrom<Instance> for KaniFunction {
+    type Error = ();
+
+    fn try_from(instance: Instance) -> Result<Self, Self::Error> {
+        let value = attributes::fn_marker(instance.def).ok_or(())?;
+        if let Ok(intrisic) = KaniIntrinsic::from_str(&value) {
+            Ok(intrisic.into())
+        } else if let Ok(model) = KaniModel::from_str(&value) {
+            Ok(model.into())
+        } else if let Ok(hook) = KaniHook::from_str(&value) {
+            Ok(hook.into())
+        } else {
+            Err(())
+        }
+    }
+}
+
+/// Find all Kani functions.
+///
+/// First try to find `kani` crate. If that exists, look for the items there.
+/// If there's no Kani crate, look for the items in `core` since we could be using `kani_core`.
+/// Note that users could have other `kani` crates, so we look in all the ones we find.
+pub fn find_kani_functions() -> HashMap<KaniFunction, FnDef> {
+    let mut kani = stable_mir::find_crates("kani");
+    if kani.is_empty() {
+        // In case we are using `kani_core`.
+        kani.extend(stable_mir::find_crates("core"));
+    }
+    debug!(?kani, "find_kani_functions");
+    let fns = kani
+        .into_iter()
+        .find_map(|krate| {
+            let kani_funcs: HashMap<_, _> = krate
+                .fn_defs()
+                .into_iter()
+                .filter_map(|fn_def| {
+                    KaniFunction::try_from(fn_def).ok().map(|kani_function| {
+                        debug!(?kani_function, ?fn_def, "Found kani function");
+                        (kani_function, fn_def)
+                    })
+                })
+                .collect();
+            // All definitions should live in the same crate, so we can return the first one.
+            // If there are no definitions, return `None` to indicate that.
+            (!kani_funcs.is_empty()).then_some(kani_funcs)
+        })
+        .unwrap_or_default();
+    if cfg!(debug_assertions) {
+        validate_kani_functions(&fns);
+    }
+    fns
+}
+
+/// Ensure we have the valid definitions.
+pub fn validate_kani_functions(kani_funcs: &HashMap<KaniFunction, FnDef>) {
+    let mut missing = 0u8;
+    for func in KaniIntrinsic::iter()
+        .map(|i| i.into())
+        .chain(KaniModel::iter().map(|m| m.into()))
+        .chain(KaniHook::iter().map(|h| h.into()))
+    {
+        if let Some(fn_def) = kani_funcs.get(&func) {
+            assert_eq!(KaniFunction::try_from(*fn_def), Ok(func), "Unexpected function marker");
```

Comment:
This is to validate the cached values. It would be wrong if we misuse the StableMIR APIs. If we keep the cache for longer than the StableMIR context, the IDs would likely be different.